File: Packages/NMP/Sources/NMP/AuthPolicy.swift
// The app-owned NIP-42 authorization surface (#8): opaque registration
// proofs for exact-instance removal, plus the completion-only policy
// callback the engine consults before signing any AUTH proof.

import NMPFFI

/// Opaque proof of one exact local-account signer installation.
///
/// The public key is presentation and activation data. Removal still requires
/// this object, so a stale cleanup cannot remove a same-key replacement.
public final class NMPAccountRegistration: @unchecked Sendable {
    public let publicKey: String
    let ffi: FfiAccountRegistration

    init(ffi: FfiAccountRegistration) {
        self.ffi = ffi
        publicKey = ffi.publicKey()
    }
}

/// Every immutable fact supplied to app-owned NIP-42 authorization policy.
public struct NMPAuthPolicyRequest: Sendable, Hashable {
    public let expectedPublicKey: String
    public let relay: String
    public let challenge: String
    public let transportGeneration: UInt64
    public let epochSequence: UInt64

    public init(
        expectedPublicKey: String,
        relay: String,
        challenge: String,
        transportGeneration: UInt64,
        epochSequence: UInt64
    ) {
        self.expectedPublicKey = expectedPublicKey
        self.relay = relay
        self.challenge = challenge
        self.transportGeneration = transportGeneration
        self.epochSequence = epochSequence
    }

    init(_ ffi: FfiAuthPolicyRequest) {
        self.init(
            expectedPublicKey: ffi.expectedPublicKey,
            relay: ffi.relay,
            challenge: ffi.challenge,
            transportGeneration: ffi.transportGeneration,
            epochSequence: ffi.epochSequence
        )
    }
}

/// The closed result vocabulary for one authorization-policy request.
public enum NMPAuthPolicyOutcome: Sendable, Equatable {
    case allow
    case deny(reason: String)
    case unavailable
    case technical(reason: String)

    var ffi: FfiAuthPolicyOutcome {
        switch self {
        case .allow: .allow
        case .deny(let reason): .deny(reason: reason)
        case .unavailable: .unavailable
        case .technical(let reason): .technical(reason: reason)
        }
    }
}

/// Exact terminal ownership failures from resolving one policy completion.
public enum NMPAuthPolicyCompletionError: Error, Sendable, Equatable {
    case alreadyCompleted
    case cancelled
    case receiverGone

    init(_ ffi: FfiAuthPolicyCompletionError) {
        switch ffi {
        case .AlreadyCompleted: self = .alreadyCompleted
        case .Cancelled: self = .cancelled
        case .ReceiverGone: self = .receiverGone
        }
    }
}

/// One-shot completion for an app-owned authorization decision.
///
/// Resolve inline for an immediate result, retain this object to resolve
/// asynchronously, or return without retaining/resolving to report
/// `.unavailable`. There is no parallel return-value policy path.
public final class NMPAuthPolicyCompletion: @unchecked Sendable {
    private let ffi: FfiAuthPolicyCompletion

    init(_ ffi: FfiAuthPolicyCompletion) {
        self.ffi = ffi
    }

    public func resolve(_ outcome: NMPAuthPolicyOutcome) throws {
        do {
            try ffi.resolve(outcome: outcome.ffi)
        } catch let error as FfiAuthPolicyCompletionError {
            throw NMPAuthPolicyCompletionError(error)
        }
    }
}

/// App-owned NIP-42 authorization policy.
///
/// `evaluate` is intentionally completion-only. `onCancelled` runs only when
/// runtime cancellation wins terminal ownership and may safely reenter NMP.
public protocol NMPAuthPolicy: AnyObject, Sendable {
    func evaluate(
        request: NMPAuthPolicyRequest,
        completion: NMPAuthPolicyCompletion
    )

    func onCancelled(request: NMPAuthPolicyRequest)
}

public extension NMPAuthPolicy {
    func onCancelled(request: NMPAuthPolicyRequest) {}
}

final class AuthPolicyBridge: FfiAuthPolicyCallback, @unchecked Sendable {
    private let policy: any NMPAuthPolicy

    init(policy: any NMPAuthPolicy) {
        self.policy = policy
    }

    func evaluate(request: FfiAuthPolicyRequest, completion: FfiAuthPolicyCompletion) {
        policy.evaluate(
            request: NMPAuthPolicyRequest(request),
            completion: NMPAuthPolicyCompletion(completion)
        )
    }

    func onCancelled(request: FfiAuthPolicyRequest) {
        policy.onCancelled(request: NMPAuthPolicyRequest(request))
    }
}

/// Opaque proof of one exact AUTH-policy installation.
public final class NMPAuthPolicyRegistration: @unchecked Sendable {
    public let publicKey: String
    let ffi: FfiAuthPolicyRegistration

    init(ffi: FfiAuthPolicyRegistration) {
        self.ffi = ffi
        publicKey = ffi.expectedPublicKey()
    }
}

extension NMPEngine {
    /// Install `policy` for one exact expected account identity.
    public func addAuthPolicy(
        expectedPublicKey: String,
        policy: any NMPAuthPolicy
    ) throws -> NMPAuthPolicyRegistration {
        let registration = try nmpRethrowing {
            try ffi.addAuthPolicy(
                expectedPublicKey: expectedPublicKey,
                callback: AuthPolicyBridge(policy: policy)
            )
        }
        return NMPAuthPolicyRegistration(ffi: registration)
    }

    /// Remove only the policy installation proven by `registration`.
    /// Repeated and stale removals return `false`.
    public func removeAuthPolicy(_ registration: NMPAuthPolicyRegistration) throws -> Bool {
        try nmpRethrowing {
            try ffi.removeAuthPolicy(registration: registration.ffi)
        }
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/AuthPolicy.kt
package com.nmp.sdk

import java.util.concurrent.atomic.AtomicBoolean
import uniffi.nmp_ffi.FfiAccountRegistration
import uniffi.nmp_ffi.FfiAuthPolicyCallback
import uniffi.nmp_ffi.FfiAuthPolicyCompletion
import uniffi.nmp_ffi.FfiAuthPolicyCompletionException
import uniffi.nmp_ffi.FfiAuthPolicyOutcome
import uniffi.nmp_ffi.FfiAuthPolicyRegistration
import uniffi.nmp_ffi.FfiAuthPolicyRequest

/** Opaque proof for one exact local-account installation. */
class NMPAccountRegistration internal constructor(
    internal val ffi: FfiAccountRegistration,
) {
    /** The installed account's hex public key. The capability instance stays opaque. */
    val publicKey: String = ffi.publicKey()
}

/** Opaque proof for one exact AUTH-policy installation. */
class NMPAuthPolicyRegistration internal constructor(
    internal val ffi: FfiAuthPolicyRegistration,
) {
    /** The account identity this exact policy installation is bound to. */
    val publicKey: String = ffi.expectedPublicKey()
}

/** Immutable facts for one exact relay AUTH challenge. */
data class NMPAuthPolicyRequest(
    val publicKey: String,
    val relay: String,
    val challenge: String,
    val transportGeneration: ULong,
    val epochSequence: ULong,
)

/** The policy decision delivered through [NMPAuthPolicyCompletion]. */
sealed interface NMPAuthPolicyOutcome {
    data object Allow : NMPAuthPolicyOutcome

    data class Deny(val reason: String) : NMPAuthPolicyOutcome

    data object Unavailable : NMPAuthPolicyOutcome

    data class Technical(val reason: String) : NMPAuthPolicyOutcome
}

/** The only terminal errors an AUTH completion can report. */
sealed class NMPAuthPolicyCompletionError(message: String) : Exception(message) {
    class AlreadyCompleted : NMPAuthPolicyCompletionError("AUTH policy request already completed")

    class Cancelled : NMPAuthPolicyCompletionError("AUTH policy request was cancelled")

    class ReceiverGone : NMPAuthPolicyCompletionError("AUTH policy receiver is gone")
}

internal interface AuthPolicyCompletionPeer {
    fun resolve(outcome: FfiAuthPolicyOutcome)

    fun close()
}

private class FfiAuthPolicyCompletionPeer(
    private val ffi: FfiAuthPolicyCompletion,
) : AuthPolicyCompletionPeer {
    override fun resolve(outcome: FfiAuthPolicyOutcome) = ffi.resolve(outcome)

    override fun close() = ffi.close()
}

/**
 * Completion-only control for one AUTH request.
 *
 * Resolve synchronously for an inline decision. To decide asynchronously, call [retain] during
 * [NMPAuthPolicy.evaluate] and keep this object alive until resolving it. If evaluation returns
 * without either resolving or retaining the completion, NMP deterministically treats the policy as
 * unavailable; this explicit handshake avoids making correctness depend on JVM garbage collection.
 */
class NMPAuthPolicyCompletion internal constructor(
    private val peer: AuthPolicyCompletionPeer,
) {
    private val retained = AtomicBoolean(false)

    /** Mark this exact completion as pending beyond the evaluation callback. */
    fun retain(): NMPAuthPolicyCompletion {
        retained.set(true)
        return this
    }

    /** Resolve this exact request once. */
    @Throws(NMPAuthPolicyCompletionError::class)
    fun resolve(outcome: NMPAuthPolicyOutcome) {
        try {
            peer.resolve(outcome.toFfi())
        } catch (_: FfiAuthPolicyCompletionException.AlreadyCompleted) {
            throw NMPAuthPolicyCompletionError.AlreadyCompleted()
        } catch (_: FfiAuthPolicyCompletionException.Cancelled) {
            throw NMPAuthPolicyCompletionError.Cancelled()
        } catch (_: FfiAuthPolicyCompletionException.ReceiverGone) {
            throw NMPAuthPolicyCompletionError.ReceiverGone()
        } catch (_: IllegalStateException) {
            // The callback scope released a completion that was neither resolved nor retained.
            throw NMPAuthPolicyCompletionError.ReceiverGone()
        }
    }

    internal fun releaseCallbackScope() {
        if (!retained.get()) peer.close()
    }
}

/** Completion-only AUTH policy. No synchronous decision is returned from [evaluate]. */
interface NMPAuthPolicy {
    fun evaluate(
        request: NMPAuthPolicyRequest,
        completion: NMPAuthPolicyCompletion,
    )

    /** Called exactly once after cancellation wins. It is safe to reenter [NMPEngine]. */
    fun onCancelled(request: NMPAuthPolicyRequest) {}
}

internal class NMPAuthPolicyBridge(
    private val policy: NMPAuthPolicy,
) : FfiAuthPolicyCallback {
    override fun evaluate(
        request: FfiAuthPolicyRequest,
        completion: FfiAuthPolicyCompletion,
    ) {
        val nativeCompletion =
            NMPAuthPolicyCompletion(FfiAuthPolicyCompletionPeer(completion))
        try {
            policy.evaluate(request.toNative(), nativeCompletion)
        } finally {
            nativeCompletion.releaseCallbackScope()
        }
    }

    override fun onCancelled(request: FfiAuthPolicyRequest) {
        policy.onCancelled(request.toNative())
    }
}

private fun FfiAuthPolicyRequest.toNative() =
    NMPAuthPolicyRequest(
        publicKey = expectedPublicKey,
        relay = relay,
        challenge = challenge,
        transportGeneration = transportGeneration,
        epochSequence = epochSequence,
    )

private fun NMPAuthPolicyOutcome.toFfi(): FfiAuthPolicyOutcome =
    when (this) {
        NMPAuthPolicyOutcome.Allow -> FfiAuthPolicyOutcome.Allow
        is NMPAuthPolicyOutcome.Deny -> FfiAuthPolicyOutcome.Deny(reason)
        NMPAuthPolicyOutcome.Unavailable -> FfiAuthPolicyOutcome.Unavailable
        is NMPAuthPolicyOutcome.Technical -> FfiAuthPolicyOutcome.Technical(reason)
    }
